File: detection/engine.go
package detection

import (
	"bytes"
	"log"
	"net"
	"strconv"
	"strings"
	"time"

	"github.com/Alexmaster12345/IntrusionShield/parser"
)

// Severity levels
const (
	SeverityLow    = 1
	SeverityMedium = 2
	SeverityHigh   = 3
)

// Alert is generated when a rule matches a packet.
type Alert struct {
	Timestamp time.Time
	RuleID    int
	Sid       int
	Severity  int
	Msg       string
	Protocol  string
	SrcIP     net.IP
	DstIP     net.IP
	SrcPort   uint16
	DstPort   uint16
	Payload   []byte
}

// Engine is the detection pipeline that matches packets against rules.
type Engine struct {
	rules   []Rule
	AlertCh chan Alert
}

// NewEngine creates an Engine loaded with built-in + file rules.
func NewEngine(rulesFile string) (*Engine, error) {
	rules := DefaultSignatures()

	if rulesFile != "" {
		fileRules, err := LoadRules(rulesFile)
		if err != nil {
			log.Printf("[detection] could not load rules file %q: %v — using built-in only", rulesFile, err)
		} else {
			rules = append(rules, fileRules...)
			log.Printf("[detection] loaded %d rules from %s", len(fileRules), rulesFile)
		}
	}

	log.Printf("[detection] engine ready with %d rules", len(rules))

	return &Engine{
		rules:   rules,
		AlertCh: make(chan Alert, 500),
	}, nil
}

// Inspect checks a packet against all rules and emits alerts.
func (e *Engine) Inspect(p *parser.Packet) {
	for _, r := range e.rules {
		if e.matches(r, p) {
			a := Alert{
				Timestamp: p.Timestamp,
				RuleID:    r.ID,
				Sid:       r.Sid,
				Severity:  r.Severity,
				Msg:       r.Msg,
				Protocol:  p.Protocol,
				SrcIP:     p.SrcIP,
				DstIP:     p.DstIP,
				SrcPort:   p.SrcPort,
				DstPort:   p.DstPort,
				Payload:   p.Payload,
			}
			select {
			case e.AlertCh <- a:
			default:
			}
		}
	}
}

func (e *Engine) matches(r Rule, p *parser.Packet) bool {
	// Protocol match
	if r.Protocol != "any" && !strings.EqualFold(r.Protocol, p.Protocol) {
		return false
	}

	// Source IP
	if r.SrcIP != "any" && !ipMatches(r.SrcIP, p.SrcIP) {
		return false
	}

	// Source port
	if r.SrcPort != "any" && r.SrcPort != "" && !portMatches(r.SrcPort, p.SrcPort) {
		return false
	}

	// Destination IP
	if r.DstIP != "any" && !ipMatches(r.DstIP, p.DstIP) {
		return false
	}

	// Destination port
	if r.DstPort != "any" && r.DstPort != "" && !portMatches(r.DstPort, p.DstPort) {
		return false
	}

	// TCP flags
	if r.Flags != "" && p.Protocol == "TCP" {
		if !flagsMatch(r.Flags, p.Flags) {
			return false
		}
	}

	// Payload content match
	if r.Content != "" {
		payload := p.Payload
		content := []byte(r.Content)
		if r.Nocase {
			payload = bytes.ToLower(payload)
			content = bytes.ToLower(content)
		}
		if !bytes.Contains(payload, content) {
			return false
		}
	}

	return true
}

func ipMatches(cidr string, ip net.IP) bool {
	if cidr == "any" || ip == nil {
		return true
	}
	if strings.Contains(cidr, "/") {
		_, network, err := net.ParseCIDR(cidr)
		if err != nil {
			return false
		}
		return network.Contains(ip)
	}
	return net.ParseIP(cidr).Equal(ip)
}

func portMatches(spec string, port uint16) bool {
	if spec == "any" || spec == "" {
		return true
	}
	// Range: "1024:65535"
	if strings.Contains(spec, ":") {
		parts := strings.SplitN(spec, ":", 2)
		lo, _ := strconv.ParseUint(parts[0], 10, 16)
		hi, _ := strconv.ParseUint(parts[1], 10, 16)
		return port >= uint16(lo) && port <= uint16(hi)
	}
	// Negation: "!80"
	if strings.HasPrefix(spec, "!") {
		n, _ := strconv.ParseUint(spec[1:], 10, 16)
		return port != uint16(n)
	}
	n, err := strconv.ParseUint(spec, 10, 16)
	if err != nil {
		return false
	}
	return port == uint16(n)
}

// flagsMatch checks TCP flags against a Snort-style flag string.
// "S"=SYN, "A"=ACK, "R"=RST, "F"=FIN, "P"=PSH, "U"=URG, "0"=no flags
func flagsMatch(spec string, f parser.TCPFlags) bool {
	if spec == "0" {
		return !f.SYN && !f.ACK && !f.RST && !f.FIN && !f.PSH && !f.URG
	}
	for _, c := range strings.ToUpper(spec) {
		switch c {
		case 'S':
			if !f.SYN {
				return false
			}
		case 'A':
			if !f.ACK {
				return false
			}
		case 'R':
			if !f.RST {
				return false
			}
		case 'F':
			if !f.FIN {
				return false
			}
		case 'P':
			if !f.PSH {
				return false
			}
		case 'U':
			if !f.URG {
				return false
			}
		}
	}
	return true
}

File: detection/engine.py
from __future__ import annotations
import ipaddress
import logging
import queue
from dataclasses import dataclass, field
from datetime import datetime
from typing import List, Optional

from detection.rules import Rule, load_rules
from detection.signatures import default_signatures
from parser.packet import Packet, TCPFlags

logger = logging.getLogger(__name__)

SEVERITY_LOW = 1
SEVERITY_MEDIUM = 2
SEVERITY_HIGH = 3


@dataclass
class Alert:
    timestamp: datetime
    rule_id: int
    sid: int
    severity: int
    msg: str
    protocol: str
    src_ip: str
    dst_ip: str
    src_port: int
    dst_port: int
    payload: bytes = field(default=b"", repr=False)


class Engine:
    def __init__(self, rules_file: str = ""):
        self.rules: List[Rule] = list(default_signatures())

        if rules_file:
            try:
                file_rules = load_rules(rules_file)
                self.rules.extend(file_rules)
                logger.info("Loaded %d rules from %s", len(file_rules), rules_file)
            except Exception as e:
                logger.warning("Could not load rules file %r: %s — using built-in only", rules_file, e)

        logger.info("Detection engine ready with %d rules", len(self.rules))
        self.alert_queue: queue.Queue[Alert] = queue.Queue(maxsize=500)

    def inspect(self, p: Packet) -> None:
        for rule in self.rules:
            if self._matches(rule, p):
                a = Alert(
                    timestamp=p.timestamp,
                    rule_id=rule.id,
                    sid=rule.sid,
                    severity=rule.severity,
                    msg=rule.msg,
                    protocol=p.protocol,
                    src_ip=p.src_ip,
                    dst_ip=p.dst_ip,
                    src_port=p.src_port,
                    dst_port=p.dst_port,
                    payload=p.payload,
                )
                try:
                    self.alert_queue.put_nowait(a)
                except queue.Full:
                    pass

    def _matches(self, r: Rule, p: Packet) -> bool:
        if r.protocol != "any" and r.protocol.upper() != p.protocol.upper():
            return False

        if r.src_ip != "any" and not _ip_matches(r.src_ip, p.src_ip):
            return False

        if r.src_port not in ("any", "") and not _port_matches(r.src_port, p.src_port):
            return False

        if r.dst_ip != "any" and not _ip_matches(r.dst_ip, p.dst_ip):
            return False

        if r.dst_port not in ("any", "") and not _port_matches(r.dst_port, p.dst_port):
            return False

        if r.flags and p.protocol.upper() == "TCP":
            if not _flags_match(r.flags, p.flags):
                return False

        if r.content:
            payload = p.payload
            content = r.content.encode()
            if r.nocase:
                payload = payload.lower()
                content = content.lower()
            if content not in payload:
                return False

        return True


def _ip_matches(spec: str, ip: str) -> bool:
    if spec == "any" or not ip:
        return True
    try:
        if "/" in spec:
            return ipaddress.ip_address(ip) in ipaddress.ip_network(spec, strict=False)
        return ipaddress.ip_address(ip) == ipaddress.ip_address(spec)
    except ValueError:
        return False


def _port_matches(spec: str, port: int) -> bool:
    if spec in ("any", ""):
        return True
    if ":" in spec:
        lo, hi = spec.split(":", 1)
        return int(lo) <= port <= int(hi)
    if spec.startswith("!"):
        return port != int(spec[1:])
    try:
        return port == int(spec)
    except ValueError:
        return False


def _flags_match(spec: str, f: TCPFlags) -> bool:
    if spec == "0":
        return not any([f.syn, f.ack, f.rst, f.fin, f.psh, f.urg])
    mapping = {"S": f.syn, "A": f.ack, "R": f.rst, "F": f.fin, "P": f.psh, "U": f.urg}
    return all(mapping.get(c, True) for c in spec.upper())
